Implement ssl context which was implemented in react socket-client 0.5.0

mvangoor opened the pull request.

elazar commented on composer.lock:
```diff
@@ -51,20 +50,20 @@
         {
             "name": "monolog/monolog",
-            "version": "1.17.2",
+            "version": "1.22.0",
```
Changes to `composer.lock` should only reflect what was changed in `composer.json` (i.e. that `react/socket-client` was upgraded).

mvangoor replied:
I see your point, composer is new to me.
I fixed this.

elazar commented on tests/ClientTest.php:
```diff
@@ -345,11 +346,12 @@ public function testAddConnectionWithSslTransportAndForceIpv4ThrowsException()
+            $this->fail('Unexpected exception was thrown');
         }
     }
+*/
```
This test should be modified to ensure that both of these options can now be enabled simultaneously.

mvangoor replied:
I am uncertain how to test this and that is why I commented it out.
The actual socket for Secure Connections is created in TcpConnecter (react/socket-client) and as far as I can see cannot be requested to check the context options.

I did find I could get the context options using:
stream_context_get_options($connection->getData('stream'))

However getData is not setup in MockConnection.

Could you help me on this one?

mvangoor replied:
Fixed

elazar commented on src/Client.php:
```diff
@@ -301,6 +323,28 @@ protected function getContext(ConnectionInterface $connection)
+        if ($this->getAllowSelfSignedFlag($connection)) {
+            $context['allow_self_signed'] = true;
+        }
+        //$context = array('socket' => $context);
```
I think removing this line, by commenting it out or otherwise, makes this code incorrect. See [these examples](http://php.net/manual/en/context.socket.php#refsect1-context.socket-examples) of using socket context options.

mvangoor replied:
The context we send to react socket-client for a secure connection is different.
As far as I can see passing the socket is not valid, I should have deleted that line and did that now.

Maybe I am wrong, but if I look at SecureConnector.php at react/socket-client it is doing a completely different thing than for a regular connection.

elazar replied:
You're correct, `SecureConnector` appears to [expect socket options specifically](https://github.com/reactphp/socket-client/blob/v0.5.3/src/SecureConnector.php#L48). However, it also appears that `TcpConnector` [does as well](https://github.com/reactphp/socket-client/blob/v0.5.3/src/TcpConnector.php#L36). As such, this PR will need to update [this line](https://github.com/mvangoor/phergie-irc-client-react/blob/71bac148f758f3c930fa0a8c32f006dbd40b9a55/src/Client.php#L321) as well to avoid breaking non-secure connections.

mvangoor replied:
Let me see if I understand you correctly here.

We should also update the regular (non-ssl) connection processing to the new 0.5.* range and send the context directly to react.
This means that functions like getSocket() will no longer be required.

If that is what you mean I will see if I can update the regular connection process aswell.

elazar replied:
At first glance, it looks like it would be possible to do so, but I think that may be better addressed in a separate PR once this one is merged.

Alternatively, we could have a PR that does nothing but update `react/socket-client` to `^0.5`, and this and the other prospective PR could be based against that one, so that the two can be developed concurrently.

mvangoor replied:
I believe you are right and we should use a seperate PR to look at socket connections in general.
I updated code to reflect discussions

elazar commented on src/Client.php:
```diff
@@ -259,6 +259,28 @@ protected function getForceIpv4Flag(ConnectionInterface $connection)
+     */
+    protected function getAllowSelfSignedFlag(ConnectionInterface $connection)
+    {
+        return (boolean) $connection->getOption('allow-selfsigned') ?: false;
```
I would use `allow-self-signed` instead of `allow-selfsigned` here, for consistency with the socket option being set.

mvangoor replied:
Changed

elazar commented on src/Client.php:
```diff
@@ -259,6 +259,28 @@ protected function getForceIpv4Flag(ConnectionInterface $connection)
+     */
+    protected function getVerifyPeerFlag(ConnectionInterface $connection)
+    {
+        return (boolean) $connection->getOption('allow-verify_peer') ?: false;
```
I would use `allow-verify-peer` instead of `allow-verify_peer` here.

mvangoor replied:
Changed

mvangoor replied:
Removed it as I cannot find a proper way of testing this and we might need to refactor socket connections completely due to new insights and possibilities of react socket-client

elazar commented:
I believe this is intended to address #25 ?

mvangoor commented:
Hi elazar,

Yes this also fixes #25, however that is not why I changed the code.
The reason is that there was no way to actually pass along any ssl context options.
With this you can do so and I added 2 extra options.

I will look at your review and update the code.

elazar commented on composer.json:
```diff
@@ -16,7 +16,7 @@
         "react/event-loop": "~0.4.0",
         "react/stream": "~0.4.2",
-        "react/socket-client": "~0.4.2",
+        "react/socket-client": "~0.5.0",
```
I'd recommend changing `~0.5.0` to `^0.5`. It appears that 0.5.3 is available, while this version restriction is only allowing 0.5.1 to be installed.

mvangoor replied:
Done

elazar commented on tests/ClientTest.php:
```diff
@@ -332,24 +332,27 @@ public function testAddConnectionWithConcreteWriteStream()
+
+        $this->client->setLogger($this->getMockLogger());
         $this->client->setResolver($this->getMockResolver());
+        $this->client->addConnection($connection);
```
I think it's sufficient for this test to simply call `addConnection()` as above. So long as an exception is no longer thrown, that's all the test should need to verify. I'm skeptical that it would be worth the effort to test for lower-level conditions beyond that.

matthewtrask replied:
Github is showing this as commented out? Is there a reason there is a commented out test in the PR?

mvangoor replied:
I have been struggling with this specific test and asked elazar to help me with it.
Once we have figured that out this will be resolved.

elazar commented:
@svpernova09 @matthewtrask I'd suggest manual testing of SSL and non-SSL connections after applying this change.

elazar commented:
@svpernova09 I believe @mvangoor started this work in order to be able to use SSL with self-signed certificates. Is there any further manual testing we should do for that use case?

svpernova09 commented:
@elazar Yes, in that case we need to find an IRC server that is using self signed cert to test this against. @mvangoor Can you share a server that is using self signed certs?

mvangoor commented:
Hi,

The reason for adding it is not the fact that (to my knowledge) any public ircd server is serving ssl without a signed cert. But rather that I am running a private (public reachible) irc server with self-signed certificate with a connection password.

I'll see if I can find a self-signed certificate ircd out there.

mvangoor commented:
My private ircd is at:
whoever.probie.nl:7000, but requires a password to connect

Found anoter one:
ircsource.baconsvin.org:6697, but the certificate has expired

Sorry to say that I cannot find another at this time

matthewtrask commented:
@mvangoor Im sitting with @svpernova09 and we are both having a hard time trying to get it up and running for testing. Can you please provide instructions for us to test properly?

matthewtrask commented on tests/ClientTest.php:
```diff
@@ -332,24 +332,27 @@ public function testAddConnectionWithConcreteWriteStream()
-            $this->assertSame(Exception::ERR_CONNECTION_STATE_UNSUPPORTED, $e->getCode());
-        }
+        print_r(stream_context_get_options($connection->getData('stream')));
+        //print_r($actual);
```
Debug code?

mvangoor replied:
I have been struggling with this specific test and asked elazar to help me with it.
Once we have figured that out this will be resolved.

elazar replied:
[I commented about this earlier](https://github.com/phergie/phergie-irc-client-react/pull/58#discussion_r99470640).

mvangoor commented:
Hi,

I have been thinking about this for a while now and how to get this moving forward.
I see 2 options currently:
- We move forward with the PR and fix all open issues. I reconfigure my private ircd temporary to allow other connections using a password so you can test.
- I close this PR completely and start fresh and look if we can expose the context externally using the Connection interface. That way if new/other context options are available people can update their code if they want.

Please let me know if you think option 1 should be focused on or option 2.

elazar commented:
@mvangoor My advice would be to focus on option 1. Connection handling is one of the larger parts of the client and should probably be refactored into a separate supporting class, but I think that would be best done in its own PR apart from the work you've already done in this one.

mvangoor commented:
Okay!

I have a test connection for you to test.
I will be removing the part about "verify_peer" as I cannot test it at this time and actually do not need it.

Once I have all the rework done I'll post a connection config so you can test it.

mvangoor commented:
Hi,

Please use this options array to create a connection.
$options = array(
  'serverHostname' => 'whoever.probie.nl',
  'serverport' => 7000,
  'password' => 'Temporary_phergie_testing',
  'username' => 'testphergie',
  'realname' => 'Phergie Client react update ^0.5',
  'nickname' => 'testphergie',
  'options' => array(
    'transport' => 'ssl',
    'allow-self-signed' => true,
  )
);

elazar commented on src/Client.php:
```diff
@@ -259,6 +259,17 @@ protected function getForceIpv4Flag(ConnectionInterface $connection)
+     */
+    protected function getAllowSelfSignedFlag(ConnectionInterface $connection)
+    {
+        return (boolean) $connection->getOption('allow-self-signed') ?: false;
```
Sorry, just noticed that Phergie configuration uses camel case (e.g. `serverHostname`) rather than this style (i.e. `allow-self-signed`). Would you mind updating this to use camel case as well?

mvangoor replied:
Not a problem, I followed the existing options as example.
See 'force-ipv4' that should have been 'forceIpv4' ?

If we go for camel case we should update them all, however this breaks api ...

just my 0.02.

elazar replied:
Fair point. Leave this as-is, then.

elazar commented:
@mvangoor The [library README](https://github.com/phergie/phergie-irc-client-react#connection-options) should probably be updated with any new configuration options.

mvangoor commented:
@elazar: Fixed the README

matthewtrask commented:
thanks @mvangoor!

mvangoor commented:
Hi 1dal,

Yes that is possible, however I have not looked at phergie in 1 year as I ended up doing other stuff.

This pull request is basically everything I needed to do to add support for self_signed option.
